src/main/java/org/opentripplanner/pricing/transit/trip/model/TransitTripDescription.java

```diff
@@ -0,0 +1,82 @@
+package org.opentripplanner.pricing.transit.trip.model;
+
+import com.google.common.collect.Range;
+import com.google.common.collect.RangeSet;
+import com.google.common.collect.TreeRangeSet;
+import lombok.Getter;
+import org.opentripplanner.model.Route;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+
+public class TransitTripDescription {
+
+    @Getter
+    private final List<TransitTripStage> tripStages = new ArrayList<>();
+
+    private final RangeSet<Integer> tripStagesTimeBounds = TreeRangeSet.create();
+
+    public TransitTripDescription(List<TransitTripStage> tripStages) {
+        if (Objects.nonNull(tripStages) && !tripStages.isEmpty()) {
+
+            TransitTripStage currentTripStage;
+            Route currentRoute = tripStages.get(0).getCurrentRoute();
+            int currentRangeFirstMinute = tripStages.get(0).getTime();
+            int currentRangeLastMinute = currentRangeFirstMinute;
+
+            for (int stageIndex = 0; stageIndex < tripStages.size(); stageIndex++) {
+                currentTripStage = tripStages.get(stageIndex);
+                this.tripStages.add(currentTripStage);
+                if (!currentRoute.getId().getId().equals(currentTripStage.getCurrentRoute().getId().getId())) {
+                    this.tripStagesTimeBounds.add(Range.closed(currentRangeFirstMinute, currentRangeLastMinute));
+                    currentRangeFirstMinute = currentTripStage.getTime();
+                } else if (stageIndex == tripStages.size() - 1) {
+                    currentRangeLastMinute = currentTripStage.getTime();
+                    this.tripStagesTimeBounds.add(Range.closed(currentRangeFirstMinute, currentRangeLastMinute));
+                }
+                currentRoute = currentTripStage.getCurrentRoute();
+                currentRangeLastMinute = currentTripStage.getTime();
+            }
+        }
+    }
+
+    public boolean isEmpty() {
+        return tripStages.isEmpty();
+    }
+
+    public boolean isTravelingAtMinute(int minute) {
+        return tripStagesTimeBounds.contains(minute);
+    }
+
+    public int getLastMinute() {
+        return tripStages.get(tripStages.size() - 1).getTime();
+    }
+
+    public int getLastMinuteOfPreviousFare(int currentMinuteOfTravel) {
+        Iterator<Range<Integer>> rangeIterator = tripStagesTimeBounds.asRanges().iterator();
```

Comment:
`asRanges` return disconnected ranges. So, if we have immediate change of bus:
`[1, 10], [10,15], [20, 24]`
it will return 
`[1,15], [20,24]`

Reply:
I think that (assuming 1-minute granularity) this case should not be a problem, because the ranges are actually only useful when we are walking from one stop to another at the analyzed minute. If a time-constrained ticket becomes invalid in the middle of a walk between two stops, then without ranges we would have to decrease time by one minute, until we reach the point, when the new ticket is actually needed. If the fare switch happens within a single minute, then in fact I do not even model this as "non-transit" part of the trip (I could do this, but I am not sure, if such precision is actually required here...).